### PR TITLE
Add session property to enforce timeout for query registration in HBO

### PR DESCRIPTION
**Pull request by feilong-liu**
## Description
HBO optimizer has several external dependency, it retrieves history statistics from external storage, get input table size information from metastore. It also canonicalize and hash the whole query plan, which can also be expensive when the query plan is large.
While latency of HBO optimizer is generally low, we have observed high latency for some queries occasionally. This has prohibited the usage of HBO optimizer in interactive workload with high latency requirement. Since HBO is an opportunistic optimization, i.e. queries still work without HBO, we can choose to skip it if it takes a long time to complete.
In this PR I add a session property to enforce timeout so that we can cap the maximum time spent in HBO optimizer

## Motivation and Context
To enable HBO for interactive workload with high latency requirement

## Impact
Enable HBO for interactive workload

## Test Plan
Run end to end with production traffic, and also add unit tests

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add session property `enforce_history_based_optimizer_register_timeout` to enforce the maximum time HBO query registration can take :pr:`23354 `
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java [hunk 1]
```diff
@@ -120,7 +125,23 @@ public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider
...
+            ExecutorService executor = Executors.newSingleThreadExecutor();
+            Future<Boolean> future = executor.submit(() -> statsCalculator.registerPlan(newPlan, session, startTimeInNano, timeoutInMilliseconds));
+            try {
+                registerSucceed = future.get(timeoutInMilliseconds, MILLISECONDS);
+            }
+            catch (Exception ignored) {
+            }
+            finally {
+                executor.shutdownNow();
+            }
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java:139]: This is the main change

**Comment by steveburnett** [comment]
* Suggest documenting the session property in https://prestodb.io/docs/current/optimizer/history-based-optimization.html#how-to-use-hbo

* Nit revision suggestion for the release note entry for consistency with the [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines): 

```
== RELEASE NOTES ==

General Changes
* Add session property `enforce_history_based_optimizer_register_timeout` to enforce the maximum time HBO query registration can take :pr:`23354 `
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for adding the doc! 

The config properties table is malformed, throwing this error in a local doc build: 

`/Users/steveburnett/Documents/GitHub/presto/presto-docs/src/main/sphinx/optimizer/history-based-optimization.rst:34: ERROR: Malformed table.  Column span alignment problem in table line 3.`

See [my comment](https://github.com/prestodb/presto/pull/23335#pullrequestreview-2210334140) in PR #23335  and my own PR #23322  for how to fix the table.

**Comment by feilong-liu** [comment]
> Thanks for adding the doc!
> 
> The config properties table is malformed, throwing this error in a local doc build:
> 
> `/Users/steveburnett/Documents/GitHub/presto/presto-docs/src/main/sphinx/optimizer/history-based-optimization.rst:34: ERROR: Malformed table. Column span alignment problem in table line 3.`
> 
> See [my comment](https://github.com/prestodb/presto/pull/23335#pullrequestreview-2210334140) in PR #23335 and my own PR #23322 for how to fix the table.

For the two PRs referenced above, I see imbalance between line 32 and line34. However in my case these two lines match. Is there any fix which can be done here?

**Comment by steveburnett** [comment]
> For the two PRs referenced above, I see imbalance between line 32 and line34. However in my case these two lines match. Is there any fix which can be done here?

I apologize, I must have been on a wrong branch from another PR. I have restarted my review confirming that I am working with your branch in this PR and you are correct, you don't have this error. Please ignore my previous comment.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pulled _this_ branch, made a new local doc build, and the new property entries in both tables look good. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java [hunk 2]
```diff
@@ -120,7 +125,23 @@ public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider
         }
 
         // Fetch and cache history based statistics of all plan nodes, so no serial network calls happen later.
-        boolean registerSucceed = statsCalculator.registerPlan(newPlan, session, startTimeInNano, timeoutInMilliseconds);
+        boolean registerSucceed = false;
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HistoricalStatisticsEquivalentPlanMarkingOptimizer.java:128]: nit: call this `registrationSucceeded`

File: presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java [hunk 3]
```diff
@@ -129,6 +130,48 @@ public void testHistoryBasedStatsCalculator()
                 anyTree(node(AggregationNode.class, node(ExchangeNode.class, anyTree(any()))).withOutputRowCount(3).withOutputSize(54)));
     }
 
+    @Test
+    public void testHistoryBasedStatsCalculatorEnforceTimeOut()
```
Comment (rschlussel) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java:134]: this looks like it tests cases where the timeout is enabled, but never exceeded. Can you add a test case where the timeout gets enforced, eg. you set the timeout really low (like to 0?) and then check that we don't use hbo stats?

Comment (feilong-liu) [reply presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java:135]: Added test with time limit to be 0

File: presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java [hunk 4]
```diff
@@ -129,6 +131,65 @@ public void testHistoryBasedStatsCalculator()
...
+                .setSystemProperty(ENFORCE_HISTORY_BASED_OPTIMIZER_REGISTRATION_TIMEOUT, "true")
+                .setSystemProperty(HISTORY_BASED_OPTIMIZER_TIMEOUT_LIMIT, "0ms")
+                .build();
+        // CBO Statistics
+        assertPlan(
+                session,
+                "SELECT * FROM nation where substr(name, 1, 1) = 'A'",
+                anyTree(node(FilterNode.class, any()).withOutputRowCount(Double.NaN)));
+
+        // HBO Statistics
```
Comment (rschlussel) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java:150]: why is this using hbo statistics? Shouldn't registration fail due to exceeding the time limit?

Comment (feilong-liu) [reply presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java]: This is to test the case where HBO statistics are written successfully (with large timeout limit), and later it tests that read of HBO statistics either succeed (if timeout limit is large) or fail (if timeout limit is 0).

I just added a few new test cases too, which is HBO statistics write failed (with 0 timeout limit), and no HBO statistics will be read.

File: presto-main/src/main/java/com/facebook/presto/testing/InMemoryHistoryBasedPlanStatisticsProvider.java [hunk 5]
```diff
@@ -80,6 +81,16 @@ public void waitProcessQueryEvents()
         }
     }
 
+    public void noProcessQueryEvents()
+    {
+        try {
+            assertFalse(semaphore.tryAcquire(10, TimeUnit.SECONDS));
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/testing/InMemoryHistoryBasedPlanStatisticsProvider.java:87]: semaphore is released during the putStats function call. If HBO failed, there will be no putStats function call, hence we will not get the semaphore here. Use this function to assert the case where HBO timeout

**Pull request merged**